from controller import Robot

# ================== Ø§Ù„Ø¥Ø¹Ø¯Ø§Ø¯ (Setup) ==================
robot = Robot()
timestep = int(robot.getBasicTimeStep())

# Ø£Ø¬Ù‡Ø²Ø© Ø§Ù„Ø±Ø¤ÙŠØ©
cam_floor = robot.getDevice("cam_floor")
if cam_floor: cam_floor.enable(timestep)

cam_reg = robot.getDevice("cam_reg")
if cam_reg:
    cam_reg.enable(timestep)
    cam_reg.recognitionEnable(timestep)

# Ø£Ø¬Ù‡Ø²Ø© Ø§Ù„Ø­Ø±ÙƒØ©
left_motor = robot.getDevice("left wheel")
right_motor = robot.getDevice("right wheel")
left_motor.setPosition(float('inf'))
right_motor.setPosition(float('inf'))

# Ø£Ø¬Ù‡Ø²Ø© Ø§Ù„Ø°Ø±Ø§Ø¹ ÙˆØ§Ù„Ù…Ù„Ù‚Ø·
arm_pitch = robot.getDevice("arm_pitch")
gripper_left = robot.getDevice("gripper_left")
gripper_right = robot.getDevice("gripper_right")

# Ù…Ø³ØªØ´Ø¹Ø±Ø§Øª Ø§Ù„Ù…Ù„Ù‚Ø·
gl_sensor = robot.getDevice("gripper_left_sensor")
gr_sensor = robot.getDevice("gripper_right_sensor")
if gl_sensor: gl_sensor.enable(timestep)
if gr_sensor: gr_sensor.enable(timestep)

def set_speeds(l, r):
    left_motor.setVelocity(l)
    right_motor.setVelocity(r)

# ================== Ù…ØªØºÙŠØ±Ø§Øª Ø§Ù„ØªØ­ÙƒÙ… ==================
state = "SCAN_FLOOR" 
colors_array = []
last_color = "GROUND"

# Ù…ØªØºÙŠØ±Ø§Øª Ø§Ù„Ù…Ù‡Ù…Ø© Ø§Ù„Ø­Ø§Ù„ÙŠØ©
task_index = 0          # Ø±Ù‚Ù… Ø§Ù„Ù…Ù‡Ù…Ø© Ø§Ù„Ø­Ø§Ù„ÙŠØ© (0 Ù„Ù„Ù…ÙƒØ¹Ø¨ Ø§Ù„Ø£ÙˆÙ„ØŒ 1 Ù„Ù„Ø«Ø§Ù†ÙŠ...)
pickup_color = ""       # Ù„ÙˆÙ† Ø§Ù„Ù…ÙƒØ¹Ø¨
delivery_color = ""     # Ù„ÙˆÙ† Ø§Ù„Ù‡Ø¯Ù

pickup_timer = 0
max_x_seen = -1.0
has_reached_peak = False

# Ù…Ø³Ø§ÙØ© Ø§Ù„ØªÙˆÙ‚Ù
STOP_DISTANCE = 0.075       # Ù„Ù„Ù…ÙƒØ¹Ø¨
STOP_DISTANCE_TARGET = 0.15 # Ù„Ù„Ù‡Ø¯Ù (Ø£Ø¨Ø¹Ø¯ Ù‚Ù„ÙŠÙ„Ø§Ù‹ Ù„ØªØ¬Ù†Ø¨ Ø§Ù„Ø§ØµØ·Ø¯Ø§Ù…)

print("ğŸš€ Pioneer 3-DX: Ù†Ø¸Ø§Ù… Ø§Ù„ØªÙ‚Ø§Ø· Ù…ØªØ³Ù„Ø³Ù„ Ø°ÙƒÙŠ Ø¬Ø§Ù‡Ø²...")

while robot.step(timestep) != -1:
    try:
        # 1) Ù…Ø³Ø­ Ø§Ù„Ø£Ø±Ø¶ÙŠØ©
        if state == "SCAN_FLOOR":
            set_speeds(2.0, 2.0)
            if cam_floor:
                img = cam_floor.getImageArray()
                if img:
                    r, g, b = img[0][0][0], img[0][0][1], img[0][0][2]
                    curr = "GROUND"
                    if r > 200 and g > 200: curr = "YELLOW"
                    elif r > 200: curr = "RED"
                    elif g > 200: curr = "GREEN"
                    elif b > 200: curr = "BLUE"
                    
                    if curr != "GROUND" and curr != last_color:
                        colors_array.append(curr)
                        print(f"âœ… ØªÙ… ØªØ³Ø¬ÙŠÙ„ Ø§Ù„Ù„ÙˆÙ†: {curr} (Ø§Ù„ØªØ±ØªÙŠØ¨ {len(colors_array)})")
                        last_color = curr
                    elif curr == "GROUND": last_color = "GROUND"

            # Ù†Ù†ØªØ¸Ø± 8 Ø£Ù„ÙˆØ§Ù† (4 Ù…ÙƒØ¹Ø¨Ø§Øª + 4 Ø£Ù‡Ø¯Ø§Ù)
            if len(colors_array) >= 8:
                state = "NEXT_TARGET"

        # 2) ØªØ¬Ù‡ÙŠØ² Ø§Ù„Ù‡Ø¯Ù Ø§Ù„ØªØ§Ù„ÙŠ
        elif state == "NEXT_TARGET":
            # Ø§Ù„ØªØ£ÙƒØ¯ Ù…Ù† ÙˆØ¬ÙˆØ¯ Ø²ÙˆØ¬ (Ù…ÙƒØ¹Ø¨ + Ù‡Ø¯Ù)
            if task_index + 4 < len(colors_array):
                pickup_color = colors_array[task_index]
                delivery_color = colors_array[task_index + 4]
                
                print(f"\nğŸ”„ Ø§Ù„Ù…Ù‡Ù…Ø© {task_index + 1}: Ù…ÙƒØ¹Ø¨ {pickup_color} -> Ù‡Ø¯Ù {delivery_color}")
                
                # ØªØµÙÙŠØ± Ù…ØªØºÙŠØ±Ø§Øª Ø§Ù„Ø¨Ø­Ø« "Ø§Ù„Ù‚Ù…Ø©"
                has_reached_peak = False
                max_x_seen = -1.0
                state = "SEARCH_CUBE"
            else:
                print("ğŸ Ø§ÙƒØªÙ…Ù„Øª Ø¬Ù…ÙŠØ¹ Ø§Ù„Ù…Ù‡Ø§Ù….")
                state = "FINISHED"

        # 3) Ø§Ù„Ø¨Ø­Ø« Ø¹Ù† Ø§Ù„Ù…ÙƒØ¹Ø¨ (Ø¨Ù†ÙØ³ Ù…Ù†Ø·Ù‚ Ø§Ù„Ù‚Ù…Ø©)
        elif state == "SEARCH_CUBE":
            objs = cam_reg.getRecognitionObjects()
            target_obj = None
            
            if objs:
                for o in objs:
                    # Ù†ØªØ¬Ø§Ù‡Ù„ Ø§Ù„ØªØ§Ø±Ø¬Øª Ø§Ù„Ø¢Ù†ØŒ Ù†Ø¨Ø­Ø« Ø¹Ù† Ø§Ù„Ù…ÙƒØ¹Ø¨ ÙÙ‚Ø·
                    if o.getModel() == "TARGET": continue

                    c = o.getColors()
                    det = "UNKNOWN"
                    if c[0] > 0.6 and c[1] < 0.4: det = "RED"
                    elif c[0] > 0.6 and c[1] > 0.6: det = "YELLOW"
                    elif c[1] > 0.6: det = "GREEN"
                    elif c[2] > 0.6: det = "BLUE"
                    
                    if det == pickup_color:
                        target_obj = o
                        break

            if target_obj:
                pos = target_obj.getPosition()
                z_dep = abs(pos[2])
                
                # --- Ø®ÙˆØ§Ø±Ø²Ù…ÙŠØ© Ø§Ù„Ø¨Ø­Ø« Ø¹Ù† Ø£Ø¹Ù„Ù‰ Ù‚ÙŠÙ…Ø© X (Ø§Ù„Ù‚Ù…Ø©) ---
                if not has_reached_peak:
                    if pos[0] >= max_x_seen:
                        max_x_seen = pos[0]
                        set_speeds(0.4, -0.4) # Ø¯ÙˆØ±Ø§Ù† Ù„Ù„ÙŠØ³Ø§Ø±
                    else:
                        has_reached_peak = True # ÙˆØ¬Ø¯Ù†Ø§ Ø§Ù„Ù‚Ù…Ø©ØŒ Ù†ØªÙˆÙ‚Ù Ø¹Ù† Ø§Ù„Ø¯ÙˆØ±Ø§Ù†
                else:
                    # Ø§Ù„ØªØ­Ø±Ùƒ Ù†Ø­Ùˆ Ø§Ù„Ù…ÙƒØ¹Ø¨
                    if z_dep > STOP_DISTANCE:
                        set_speeds(0.8, 0.8)
                    else:
                        set_speeds(0, 0)
                        state = "PICKUP_ACTION"
                        pickup_timer = 0
            else:
                set_speeds(0.5, -0.5)

        # 4) Ø¹Ù…Ù„ÙŠØ© Ø§Ù„Ø§Ù„ØªÙ‚Ø§Ø· (ÙØªØ­ -> Ù†Ø²ÙˆÙ„ -> Ø¥ØºÙ„Ø§Ù‚)
        elif state == "PICKUP_ACTION":
            pickup_timer += 1
            
            # ÙØªØ­ Ø§Ù„Ø£ØµØ§Ø¨Ø¹ ÙˆÙ‡ÙŠ ÙÙŠ Ø§Ù„Ø£Ø¹Ù„Ù‰
            if pickup_timer == 1:
                gripper_left.setPosition(0.09) 
                gripper_right.setPosition(0.09)
                print("ğŸ‘ ÙØªØ­ Ø§Ù„Ø£ØµØ§Ø¨Ø¹...")

            # Ù†Ø²ÙˆÙ„ Ø§Ù„Ø°Ø±Ø§Ø¹
            elif pickup_timer == 40:
                arm_pitch.setPosition(0.8)
                print("ğŸ”½ Ù†Ø²ÙˆÙ„ Ø§Ù„Ø°Ø±Ø§Ø¹...")

            # Ø¥ØºÙ„Ø§Ù‚ Ø§Ù„Ø£ØµØ§Ø¨Ø¹
            elif pickup_timer == 90:
                gripper_left.setPosition(0.0) 
                gripper_right.setPosition(0.0)
                print("âœŠ Ø¥ØºÙ„Ø§Ù‚ Ø§Ù„Ø£ØµØ§Ø¨Ø¹...")

            # Ø§Ù„ØªØ­Ù‚Ù‚ Ù…Ù† Ø§Ù„Ù…Ø³ØªØ´Ø¹Ø±Ø§Øª
            elif pickup_timer == 150:
                if gl_sensor.getValue() > 0.002 or gr_sensor.getValue() > 0.002:
                    print(f"ğŸ’ ØªÙ… Ø§Ù„Ø¥Ù…Ø³Ø§Ùƒ Ø¨Ù€ {pickup_color}")
                    state = "LIFTING"
                    pickup_timer = 0
                else:
                    print("âš ï¸ ÙØ´Ù„ Ø§Ù„Ø¥Ù…Ø³Ø§ÙƒØŒ Ø¥Ø¹Ø§Ø¯Ø© Ø§Ù„Ù…Ø­Ø§ÙˆÙ„Ø©...")
                    state = "RETRY_MOVE"
                    pickup_timer = 0

        # 5) Ø§Ù„ØªØ±Ø§Ø¬Ø¹ ÙˆØ¥Ø¹Ø§Ø¯Ø© Ø§Ù„ØªÙ…Ø±ÙƒØ² (Retry Logic)
        elif state == "RETRY_MOVE":
            pickup_timer += 1
            arm_pitch.setPosition(0.0) # Ø±ÙØ¹ Ø§Ù„Ø°Ø±Ø§Ø¹ ÙÙˆØ±Ø§Ù‹
            
            if pickup_timer < 50:
                set_speeds(-1.2, -1.2) # Ø±Ø¬ÙˆØ¹ Ù„Ù„Ø®Ù„Ù
            else:
                # ØªØµÙÙŠØ± Ø§Ù„Ù…ØªØºÙŠØ±Ø§Øª Ù„Ø¥Ø¹Ø§Ø¯Ø© Ø§Ù„Ø¨Ø­Ø« Ø§Ù„Ø¯Ù‚ÙŠÙ‚
                print("ğŸ”„ Ø¥Ø¹Ø§Ø¯Ø© Ø§Ù„Ø¨Ø­Ø« Ø¹Ù† Ø§Ù„Ù…ÙƒØ¹Ø¨...")
                has_reached_peak = False # Ø¥Ø¹Ø§Ø¯Ø© ØªÙØ¹ÙŠÙ„ Ù…Ù†Ø·Ù‚ Ø§Ù„Ù‚Ù…Ø©
                max_x_seen = -1.0
                set_speeds(0, 0)
                state = "SEARCH_CUBE" # Ø§Ù„Ø¹ÙˆØ¯Ø© Ù„Ø­Ø§Ù„Ø© Ø§Ù„Ø¨Ø­Ø«
                pickup_timer = 0

        # 6) Ø§Ù„Ø±ÙØ¹ ÙˆØ§Ù„ØªÙˆØ¬Ù‡ Ù„Ù„Ø¨Ø­Ø« Ø¹Ù† Ø§Ù„ØªØ§Ø±Ø¬Øª
        elif state == "LIFTING":
            arm_pitch.setVelocity(0.5) # Ø³Ø±Ø¹Ø© Ù‡Ø§Ø¯Ø¦Ø©
            arm_pitch.setPosition(-0.5)
            
            pickup_timer += 1
            if pickup_timer > 100:
                arm_pitch.setVelocity(1.0)
                
                # Ø§Ù„Ø§Ù†ØªÙ‚Ø§Ù„ Ù„Ù„Ø¨Ø­Ø« Ø¹Ù† Ø§Ù„Ù‡Ø¯Ù (Target)
                print(f"ğŸ” Ø§Ù„Ø¨Ø­Ø« Ø¹Ù† Ø§Ù„Ù‡Ø¯Ù: {delivery_color}")
                # ØªØµÙÙŠØ± Ù…ØªØºÙŠØ±Ø§Øª Ø§Ù„Ø¨Ø­Ø« Ù…Ø±Ø© Ø£Ø®Ø±Ù‰ Ù„Ø§Ø³ØªØ®Ø¯Ø§Ù…Ù‡Ø§ Ù…Ø¹ Ø§Ù„ØªØ§Ø±Ø¬Øª
                has_reached_peak = False
                max_x_seen = -1.0
                state = "SEARCH_TARGET"
                pickup_timer = 0

        # 7) Ø§Ù„Ø¨Ø­Ø« Ø¹Ù† Ø§Ù„Ù‡Ø¯Ù (TARGET) - Ø¥Ø¶Ø§ÙØ© Ø¬Ø¯ÙŠØ¯Ø© Ø¨Ù†ÙØ³ Ø§Ù„Ù…Ù†Ø·Ù‚
        elif state == "SEARCH_TARGET":
            objs = cam_reg.getRecognitionObjects()
            target_obj = None
            
            if objs:
                for o in objs:
                    # Ø§Ù„Ø´Ø±Ø·: Ø§Ù„Ù…ÙˆØ¯ÙŠÙ„ ÙŠØ¬Ø¨ Ø£Ù† ÙŠÙƒÙˆÙ† TARGET ÙˆØ§Ù„Ù„ÙˆÙ† Ù…Ø·Ø§Ø¨Ù‚ Ù„Ù„Ù‡Ø¯Ù
                    if o.getModel() == "TARGET":
                        c = o.getColors()
                        det = "UNKNOWN"
                        if c[0] > 0.6 and c[1] < 0.4: det = "RED"
                        elif c[0] > 0.6 and c[1] > 0.6: det = "YELLOW"
                        elif c[1] > 0.6: det = "GREEN"
                        elif c[2] > 0.6: det = "BLUE"
                        
                        if det == delivery_color:
                            target_obj = o
                            break
            
            if target_obj:
                pos = target_obj.getPosition()
                z_dep = abs(pos[2])
                
                # --- Ù†ÙØ³ Ø®ÙˆØ§Ø±Ø²Ù…ÙŠØ© Ø§Ù„Ù‚Ù…Ø© (Peak Search) ---
                if not has_reached_peak:
                    if pos[0] >= max_x_seen:
                        max_x_seen = pos[0]
                        set_speeds(0.4, -0.4)
                    else:
                        has_reached_peak = True
                else:
                    # Ø§Ù„ØªÙˆØ¬Ù‡ Ù†Ø­Ùˆ Ø§Ù„Ù‡Ø¯Ù
                    if z_dep > STOP_DISTANCE_TARGET:
                        set_speeds(0.8, 0.8)
                    else:
                        set_speeds(0, 0)
                        state = "DROP_ACTION" # Ø­Ø§Ù„Ø© ÙˆØ¶Ø¹ Ø§Ù„Ù…ÙƒØ¹Ø¨
                        pickup_timer = 0
            else:
                set_speeds(0.5, -0.5)

        # 8) ÙˆØ¶Ø¹ Ø§Ù„Ù…ÙƒØ¹Ø¨ ÙˆØ§Ù„Ø§Ù†ØªÙ‡Ø§Ø¡
        elif state == "DROP_ACTION":
            pickup_timer += 1
            
            if pickup_timer == 10:
                print("â¬ ÙˆØ¶Ø¹ Ø§Ù„Ù…ÙƒØ¹Ø¨...")
                gripper_left.setPosition(0.09)
                gripper_right.setPosition(0.09)
            
            elif pickup_timer == 50:
                set_speeds(-1.0, -1.0) # Ø§Ø¨ØªØ¹Ø§Ø¯
                
            elif pickup_timer == 100:
                set_speeds(0, 0)
                task_index += 1 # Ø§Ù„Ø§Ù†ØªÙ‚Ø§Ù„ Ù„Ù„Ù…Ù‡Ù…Ø© Ø§Ù„ØªØ§Ù„ÙŠØ©
                state = "NEXT_TARGET"
                pickup_timer = 0

        elif state == "FINISHED":
            set_speeds(0, 0)

    except Exception as e:
        print(f"âš ï¸ Ø®Ø·Ø£: {e}")
        break